### PR TITLE
First pass rough draft of Queue excercise

### PR DESCRIPTION
**Pull request by lorecrafting**
w00t

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,25 @@
-# js-queue
-Queue DataStructure exercise in es6
+# The Queue Datastructure
+
+We will be creating a **module** that helps us generate a **Queue**. A queue is a data structure in which the first element is inserted from one end called REAR (also called tail), and the deletion of existing element takes place from the other end called as FRONT (also called head).  This makes queue a FIFO (first in first out), or the element that is inserted first will always be the first to be removed.
```
Comment (theRemix) [inline README.md]: explicitly say that the goal is to create an **es6 Class** to be run in the node environment

**Review by theRemix — CHANGES_REQUESTED** [review]

File: lib/Queue.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/**
+ * Queue => Stack
+ * -----------------------------
+ * Queue a stack data structure.
+ *
```
Comment (theRemix) [inline lib/Queue.js:5]: describe that there should be a constructor method, and the constructor arguments (_none_)

File: lib/Queue.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+/**
+ * Queue => Stack
+ * -----------------------------
+ * Queue a stack data structure.
```
Comment (theRemix) [inline lib/Queue.js]: "Queue data structure"

_a Queue is not a Stack_

File: lib/Queue.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/**
+ * Queue => Stack
```
Comment (theRemix) [inline lib/Queue.js]: "Queue Class"

_a Queue is not a Stack_

**Review by theRemix — CHANGES_REQUESTED** [review]
queue datastructure is not a stack

File: package.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "js-queue",
+  "version": "0.0.1",
+  "description": "A queue datastructure excercise implemented in javascript.",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha --watch"
```
Comment (theRemix) [inline package.json]: add a `bail` flag

**Review by theRemix — CHANGES_REQUESTED** [review]

File: lib/Queue.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+   * @return {Number}
+   */
+
+  /**
+   * Queue Class Method => enqueue(item)
+   * -----------------------------
+   * Puts an item in the queue.
+   *
+   * @name enqueue
+   * @param item
```
Comment (theRemix) [inline lib/Queue.js:39]: describe what enqueue should return (nothing? or new _size?)

**Review by theRemix — CHANGES_REQUESTED** [review]
enqueue

File: README.md [hunk 7]
```diff
@@ -1,2 +1,25 @@
...
+
```
Comment (theRemix) [inline README.md:4]: Explain that they should will be using an anonymous `Object` as an internal datastore type, cannot use `Array`.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,25 @@
-# js-queue
-Queue DataStructure exercise in es6
+# The Queue Datastructure
```
Comment (theRemix) [inline README.md:1]: "Data Structure"
two words

**Review by theRemix — CHANGES_REQUESTED** [review]
readme updates

File: test/test.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    });
+    it('should have a dequeue() function', function() {
+      myQueue.dequeue.should.be.a('function');
+    });
+    it('should have a peek() function', function() {
+      myQueue.peek.should.be.a('function');
+    });
+  });
+
+  describe('Queue behavior', function() {
```
Comment (ghost) [inline test/test.js]: Currently with this title the output is:

`1) Queue Queue behavior A new queue should start with a size of 0:`

consider removing "Queue" from this describe block so the output is instead:

`1) Queue behavior A new queue should start with a size of 0:`

File: README.md [hunk 10]
```diff
@@ -1,2 +1,25 @@
...
+
+We will be creating a **module** that helps us generate a **Queue**. A queue is a data structure in which the first element is inserted from one end called REAR (also called tail), and the deletion of existing element takes place from the other end called as FRONT (also called head).  This makes queue a FIFO (first in first out), or the element that is inserted first will always be the first to be removed.
+
+## Methods
+
+### size()
+Returns the size of the stack
+
+### push(item)
```
Comment (ghost) [inline README.md]: `push` should be `enqueue` instead?

File: README.md [hunk 11]
```diff
@@ -1,2 +1,25 @@
...
+Adds an item to the top of the stack structure
+
+### pop()
```
Comment (ghost) [inline README.md:13]: `pop` should be `dequeue` instead?

File: lib/Queue.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+ * Queue a stack data structure.
+ *
+ * A queue is a data structure in which the first element is inserted from one end called
+ * REAR (also called tail), and the deletion of existing element takes place from the other
+ * end called as FRONT (also called head).  This makes queue a FIFO (first in first out), or
+ * the element that is inserted first will always be the first to be removed.
+ *
+ * @private {object} _storage              Default value should be an empty object.
+ * @private {number} _start                Default value should be set 0.
+ * @private {number} _end                  Default value should be set 0.
```
Comment (theRemix) [inline lib/Queue.js]: definition to allow different implementation types

**Review by theRemix — CHANGES_REQUESTED** [review]
is there a way to redefine the instructions such that a student can implement Queue in other ways?

consider these two branches
**remix_complete_1**
https://github.com/theRemix/js-queue/blob/remix_complete_1/lib/Queue.js
passes all tests using the names and implementation details defined exactly as the comments suggest. (how you probably solved it too)

**remix_complete_2**
https://github.com/theRemix/js-queue/blob/remix_complete_2/lib/Queue.js
passes all tests using a different implementation, i updated the comments describe the new version.

File: test/test.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+      myQueue.enqueue('b');
+      myQueue.dequeue();
+      myQueue.dequeue().should.equal('b');
+    });
+    it('Peek should look at the first item in line without removing it', function() {
+      myQueue.enqueue('a');
+      myQueue.enqueue('b');
+      myQueue.peek().should.equal('a');
+      myQueue.dequeue().should.equal('a');
+    });
```
Comment (ghost) [inline test/test.js:81]: What should the `peak` method return if there are 0 items in the queue? I'm thinking a return value of `null` would be good. thoughts?

Comment (theRemix) [reply test/test.js:81]: i think `null` would be good.

File: README.md (hunk 1 above)
Comment (lorecrafting) [reply README.md]: ✅

File: lib/Queue.js (hunk 4 above)
Comment (lorecrafting) [reply lib/Queue.js]: ✅

File: lib/Queue.js (hunk 3 above)
Comment (lorecrafting) [reply lib/Queue.js]: ✅

File: lib/Queue.js (hunk 2 above)
Comment (lorecrafting) [reply lib/Queue.js:5]: ✅

File: package.json (hunk 5 above)
Comment (lorecrafting) [reply package.json]: ✅

File: lib/Queue.js (hunk 6 above)
Comment (lorecrafting) [reply lib/Queue.js:39]: good point, having it return new size

File: README.md (hunk 8 above)
Comment (lorecrafting) [reply README.md:1]: ✅

File: README.md (hunk 7 above)
Comment (lorecrafting) [reply README.md:4]: ✅

File: test/test.js (hunk 9 above)
Comment (lorecrafting) [reply test/test.js]: ✅

File: README.md (hunk 10 above)
Comment (lorecrafting) [reply README.md]: ✅

File: README.md (hunk 11 above)
Comment (lorecrafting) [reply README.md:13]: ✅

File: test/test.js (hunk 13 above)
Comment (lorecrafting) [reply test/test.js:81]: ✅

File: lib/Queue.js (hunk 12 above)
Comment (lorecrafting) [reply lib/Queue.js]: Cool second implementation! Updated the exercise prompt to say that the datastore object is optional and there is a way to finish the exercise without it.

**Pull request merged**